psydac/api/fem.py

```diff
@@ -215,31 +219,34 @@ def __init__(self, expr, kernel_expr, *args, **kwargs):
         assert np.array_equal(quad_order, get_quad_order(self.spaces[1]))
         BasicDiscrete.__init__(self, expr, kernel_expr, quad_order=quad_order, **kwargs)
         #...
-        grid              = QuadratureGrid( test_space, axis, test_ext, trial_space=trial_space)
-        self._grid        = grid
-        self._test_basis  = BasisValues( test_space,  nderiv = self.max_nderiv , trial=False, grid=grid, ext=test_ext)
-        self._trial_basis = BasisValues( trial_space, nderiv = self.max_nderiv , trial=True, grid=grid, ext=trial_ext)
+        grid                = QuadratureGrid( test_space, axis, test_ext, trial_space=trial_space)
+        self._grid          = grid
+        self._test_basis    = BasisValues( test_space,  nderiv = self.max_nderiv , trial=False, grid=grid, ext=test_ext)
+        self._trial_basis   = BasisValues( trial_space, nderiv = self.max_nderiv , trial=True, grid=grid, ext=trial_ext)
 
+        if isinstance(test_space.vector_space, BlockVectorSpace):
+            vector_space = test_space.vector_space.spaces[0]
+            if isinstance(test_space.vector_space, BlockVectorSpace):
+                vector_space = test_space.vector_space.spaces[0]
+        else:
+            vector_space = test_space.vector_space
+
+        starts = vector_space.starts
+        ends   = vector_space.ends
+        npts   = vector_space.npts
+
+        self._starts = tuple(0 if i==0   else 1 for i in starts)
+        self._ends   = tuple(0 if i+1==n else 1 for i,n in zip(ends, npts))
```

Comment:
Why are they not booleans then?

Reply:
because, I use them in the assembly algorithm to calculate the bounds of the element matrix when we are close to the boundary

Reply:
In the `linalg` part of Psydac the attributes `starts` and `ends` have a very different meaning than here, so I suggest to change the name of these variables.

Also, I suggest to use `bool` objects and then convert them to integers when needed by writing `int(var)`.

Reply:
we always need them as integers, that's why I'm storing them like that, as they will be passed as arguments , otherwise we will need to convert them in each function call

Reply:
Isn't the arguments' list created as a unique flattened list at some point in the process? You can convert them there, not at each function call.

Reply:
okay, I will do that

Reply:
fixed